helix-term/src/ui/editor.rs

```diff
@@ -735,8 +742,35 @@ impl EditorView {
                 // first execute whatever put us into insert mode
                 self.last_insert.0.execute(cxt);
                 // then replay the inputs
-                for &key in &self.last_insert.1.clone() {
-                    self.insert_mode(cxt, key)
+                for key in self.last_insert.1.clone() {
+                    match key {
+                        InsertEvent::Key(key) => self.insert_mode(cxt, key),
+                        InsertEvent::CompletionApply(compl) => {
+                            let (view, doc) = current!(cxt.editor);
+
+                            doc.restore(view.id);
+
+                            let text = doc.text().slice(..);
+                            let cursor = doc.selection(view.id).primary().cursor(text);
+
+                            log::error!("Replaying completion: {compl:?}");
+
+                            let shift_position =
+                                |pos: usize| -> usize { pos + cursor - compl.trigger_pos };
+
+                            let tx = Transaction::change(
+                                doc.text(),
+                                compl.transaction.changes_iter().map(|(start, end, t)| {
+                                    (shift_position(start), shift_position(end), t)
+                                }),
```

Comment:
This will can include multiple changes like new imports being added. You will want to find and use only the single change where trigger position was between start and end.

Reply:
Changed that. The only thing that bothers me is doing this `filter_map` on every replay. Maybe there is some way to cache that. On the other hand this code does not run when replay is not used.
I didn't noticed any lag when replaying on the release build though, so maybe such micro optimization is not worth it.
@archseer what do you think ?